spring/src/main/java/org/axonframework/spring/serialization/avro/SpecificRecordBaseClasspathAvroSchemaLoader.java

```diff
@@ -0,0 +1,79 @@
+ * limitations under the License.
+ */
+
+package org.axonframework.spring.serialization.avro;
+
+import org.apache.avro.Schema;
+import org.apache.avro.specific.SpecificRecordBase;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.core.type.filter.AssignableTypeFilter;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Avro schema loader loading schemas embedded into Java classes generated by Avro Java Maven generator, which include
+ * the original schema in a static field.
+ */
+public class SpecificRecordBaseClasspathAvroSchemaLoader implements ClasspathAvroSchemaLoader {
+
+    private final ClassPathScanningCandidateComponentProvider candidateProvider;
+
+    public SpecificRecordBaseClasspathAvroSchemaLoader(ResourceLoader resourceLoader) {
+        candidateProvider = new ClassPathScanningCandidateComponentProvider(false);
+        candidateProvider.setResourceLoader(resourceLoader);
+        candidateProvider.addIncludeFilter(new AssignableTypeFilter(SpecificRecordBase.class));
+    }
+
+
+    @Override
+    public List<Schema> load(List<String> packageNames) {
+        return packageNames.stream()
+                           .map(this::scan)
+                           .flatMap(List::stream)
+                           .collect(Collectors.toList());
+    }
+
+    private List<Schema> scan(String packageName) {
+        Set<BeanDefinition> candidates = candidateProvider.findCandidateComponents(packageName);
+        return candidates.stream()
+                         .map(candidate -> {
+                             try {
+                                 Class<?> clazz = Class.forName(candidate.getBeanClassName());
+                                 if (SpecificRecordBase.class.isAssignableFrom(clazz)) {
+                                     @SuppressWarnings("unchecked")
+                                     Class<SpecificRecordBase> specificRecordBaseClass = (Class<SpecificRecordBase>) clazz;
+                                     return getClassSchema(specificRecordBaseClass);
+                                 } else {
+                                     return null;
+                                 }
+                             } catch (Exception e) {
+                                 return null;
+                             }
+                         }).filter(Objects::nonNull)
+                         .collect(Collectors.toList());
+    }
+
+    private static Schema getClassSchema(Class<SpecificRecordBase> clazz)
+            throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+        return (Schema) clazz.getMethod("getClassSchema").invoke(null); // invoke static method
```

Comment:
So, this approach might fail if, for whatever reason, the scanned-for packages contain classes that have a `static getClassSchema` method that does not align with what we expect, right?

Reply:
This is an implicit contract: All classes generated for java inherit from SpecificRecordBase and have default static methods generated. So you could theoretically write your own class breaking this contract, but is is highly unlikely and discouraged and then it could break, but that's ok. We do try/catch around this call and would just not load such misused classes.